File: apps/api/app/api/routes/calendar.py
"""Authenticated calendar routes."""

from __future__ import annotations

from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from sqlalchemy.orm import Session as OrmSession

from app.db.models import CalendarCategory, CalendarEvent
from app.db.session import get_db
from app.modules.calendar.schemas import CategoryCreate, CategoryResponse, EventCreate, EventListResponse, EventReminderInput, EventReminderResponse, EventResponse, EventUpdate
from app.modules.calendar.service import add_reminder, create_category, create_event, delete_category, delete_event, delete_reminder, get_event, list_categories, list_events, update_event, update_reminder
from app.modules.identity.dependencies import AuthContext, get_auth_context, require_csrf, require_permission

router = APIRouter(prefix="/api/v1", tags=["calendar"])


def _category_response(item: CalendarCategory) -> CategoryResponse:
    return CategoryResponse(id=item.id, name=item.name, color=item.color)


def _event_response(event: CalendarEvent) -> EventResponse:
    return EventResponse(id=event.id, title=event.title, description=event.description, location=event.location, starts_at=event.starts_at, ends_at=event.ends_at, all_day=event.all_day, category=_category_response(event.category) if event.category else None, created_at=event.created_at, updated_at=event.updated_at, reminders=[EventReminderResponse(id=item.id, scheduled_for=item.scheduled_for, offset_minutes=item.offset_minutes, status=item.status, delivered_at=item.delivered_at) for item in event.reminders])


def _parse_from(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise HTTPException(status_code=422, detail="from must be an ISO-8601 timestamp") from exc
    return parsed.astimezone(UTC) if parsed.tzinfo else parsed.replace(tzinfo=UTC)


@router.get("/calendar/events", response_model=EventListResponse)
def list_all(start_from: str | None = Query(default=None, alias="from"), start_to: str | None = Query(default=None, alias="to"), category: str | None = None, limit: int = 100, cursor: str | None = None, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventListResponse:
    require_permission("calendar.read", context)
    items = list_events(db, context.user.id, start_from=_parse_from(start_from), start_to=_parse_from(start_to), category=category, limit=limit, cursor=cursor)
    return EventListResponse(items=[_event_response(item) for item in items], next_cursor=items[-1].id if len(items) == min(max(limit, 1), 200) else None)


@router.post("/calendar/events", response_model=EventResponse, status_code=status.HTTP_201_CREATED)
def create(payload: EventCreate, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventResponse:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    try:
        return _event_response(create_event(db, context.user.id, payload, request.headers.get("Idempotency-Key")))
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc


@router.get("/calendar/events/{event_id}", response_model=EventResponse)
def get_one(event_id: str, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventResponse:
    require_permission("calendar.read", context)
    event = get_event(db, context.user.id, event_id)
    if event is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")
    return _event_response(event)


@router.patch("/calendar/events/{event_id}", response_model=EventResponse)
def update(event_id: str, payload: EventUpdate, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventResponse:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    try:
        event = update_event(db, context.user.id, event_id, payload, request.headers.get("Idempotency-Key"))
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    if event is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")
    return _event_response(event)


@router.delete("/calendar/events/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete(event_id: str, request: Request, response: Response, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> None:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    if delete_event(db, context.user.id, event_id, request.headers.get("Idempotency-Key")) is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")


@router.post("/calendar/events/{event_id}/reminders", response_model=EventResponse)
def add_event_reminder(event_id: str, payload: EventReminderInput, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventResponse:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    try:
        event = add_reminder(db, context.user.id, event_id, payload, request.headers.get("Idempotency-Key"))
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    if event is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")
    return _event_response(event)


@router.patch("/calendar/reminders/{reminder_id}", response_model=EventResponse)
def update_event_reminder(reminder_id: str, payload: EventReminderInput, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> EventResponse:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    try:
        event = update_reminder(db, context.user.id, reminder_id, payload, request.headers.get("Idempotency-Key"))
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    if event is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reminder not found")
    return _event_response(event)


@router.delete("/calendar/reminders/{reminder_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event_reminder(reminder_id: str, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> None:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    if not delete_reminder(db, context.user.id, reminder_id, request.headers.get("Idempotency-Key")):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reminder not found")


@router.get("/calendar/categories", response_model=list[CategoryResponse])
def categories(db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> list[CategoryResponse]:
    require_permission("calendar.read", context)
    return [_category_response(item) for item in list_categories(db, context.user.id)]


@router.post("/calendar/categories", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
def category_create(payload: CategoryCreate, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> CategoryResponse:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    try:
        return _category_response(create_category(db, context.user.id, payload, request.headers.get("Idempotency-Key")))
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc


@router.delete("/calendar/categories/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def category_delete(category_id: str, request: Request, db: OrmSession = Depends(get_db), context: AuthContext = Depends(get_auth_context)) -> None:
    require_csrf(request, context)
    require_permission("calendar.write", context)
    if not delete_category(db, context.user.id, category_id, request.headers.get("Idempotency-Key")):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")

File: apps/web/lib/calendar.ts
import { authenticatedFetch } from "./auth";

export type CalendarCategory = {
  id: string;
  name: string;
  color: string | null;
};

export type CalendarReminder = {
  id: string;
  scheduled_for: string;
  offset_minutes: number | null;
  status: string;
  delivered_at: string | null;
};

export type CalendarEvent = {
  id: string;
  title: string;
  description: string | null;
  location: string | null;
  starts_at: string;
  ends_at: string;
  all_day: boolean;
  category: CalendarCategory | null;
  created_at: string;
  updated_at: string;
  reminders: CalendarReminder[];
};

export type EventInput = {
  title: string;
  description?: string | null;
  location?: string | null;
  starts_at: string;
  ends_at: string;
  all_day?: boolean;
  category?: string | null;
  reminders?: Array<{ scheduled_for?: string; offset_minutes?: number }>;
};

export type ReminderInput = {
  scheduled_for?: string;
  offset_minutes?: number;
};

async function parse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const body = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(body?.detail ?? `Calendar request failed with ${response.status}`);
  }
  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}

function idempotencyKey(): string {
  return typeof crypto.randomUUID === "function" ? crypto.randomUUID() : `${Date.now()}-${Math.random()}`;
}

export async function listEvents(from?: string, to?: string, category?: string): Promise<CalendarEvent[]> {
  const params = new URLSearchParams();
  if (from) params.set("from", from);
  if (to) params.set("to", to);
  if (category) params.set("category", category);
  const query = params.toString();
  const response = await authenticatedFetch(`/api/v1/calendar/events${query ? `?${query}` : ""}`, { cache: "no-store" });
  const body = await parse<{ items: CalendarEvent[] }>(response);
  return body.items;
}

export async function createEvent(input: EventInput): Promise<CalendarEvent> {
  return parse<CalendarEvent>(await authenticatedFetch("/api/v1/calendar/events", { method: "POST", headers: { "Content-Type": "application/json", "Idempotency-Key": idempotencyKey() }, body: JSON.stringify(input) }));
}

export async function updateEvent(id: string, input: Partial<EventInput>): Promise<CalendarEvent> {
  return parse<CalendarEvent>(await authenticatedFetch(`/api/v1/calendar/events/${encodeURIComponent(id)}`, { method: "PATCH", headers: { "Content-Type": "application/json", "Idempotency-Key": idempotencyKey() }, body: JSON.stringify(input) }));
}

export async function deleteEvent(id: string): Promise<void> {
  await parse<void>(await authenticatedFetch(`/api/v1/calendar/events/${encodeURIComponent(id)}`, { method: "DELETE", headers: { "Idempotency-Key": idempotencyKey() } }));
}

export async function addEventReminder(eventId: string, input: ReminderInput): Promise<CalendarEvent> {
  return parse<CalendarEvent>(await authenticatedFetch(`/api/v1/calendar/events/${encodeURIComponent(eventId)}/reminders`, { method: "POST", headers: { "Content-Type": "application/json", "Idempotency-Key": idempotencyKey() }, body: JSON.stringify(input) }));
}

export async function updateReminder(reminderId: string, input: ReminderInput): Promise<CalendarEvent> {
  return parse<CalendarEvent>(await authenticatedFetch(`/api/v1/calendar/reminders/${encodeURIComponent(reminderId)}`, { method: "PATCH", headers: { "Content-Type": "application/json", "Idempotency-Key": idempotencyKey() }, body: JSON.stringify(input) }));
}

export async function deleteReminder(reminderId: string): Promise<void> {
  await parse<void>(await authenticatedFetch(`/api/v1/calendar/reminders/${encodeURIComponent(reminderId)}`, { method: "DELETE", headers: { "Idempotency-Key": idempotencyKey() } }));
}

export async function listCategories(): Promise<CalendarCategory[]> {
  const response = await authenticatedFetch("/api/v1/calendar/categories", { cache: "no-store" });
  return parse<CalendarCategory[]>(response);
}

export async function createCategory(name: string, color?: string): Promise<CalendarCategory> {
  return parse<CalendarCategory>(await authenticatedFetch("/api/v1/calendar/categories", { method: "POST", headers: { "Content-Type": "application/json", "Idempotency-Key": idempotencyKey() }, body: JSON.stringify({ name, color: color || null }) }));
}

export async function deleteCategory(id: string): Promise<void> {
  await parse<void>(await authenticatedFetch(`/api/v1/calendar/categories/${encodeURIComponent(id)}`, { method: "DELETE", headers: { "Idempotency-Key": idempotencyKey() } }));
}
